src/ray/raylet/scheduling_resources.cc

```diff
@@ -135,6 +135,206 @@ const std::unordered_map<std::string, double> &ResourceSet::GetResourceMap() con
+    RAY_CHECK(it != resource_availabilities_.end());
+
+    auto &whole_ids = it->second.first;
+    auto &fractional_ids = it->second.second;
+
+    if (resource_quantity >= 1) {
+      // Handle the whole case.
+      int64_t whole_quantity = resource_quantity;
+      RAY_CHECK(whole_quantity == resource_quantity);
+      RAY_CHECK(whole_ids.size() >= whole_quantity);
+
+      for (int64_t i = 0; i < whole_quantity; ++i) {
+        resource_ids.push_back(std::make_pair(whole_ids.back(), 1));
+        whole_ids.pop_back();
+      }
+
+    } else {
+      // Handle the fractional case.
+      bool found = false;
+      for (auto &fractional_id : fractional_ids) {
+        int64_t resource_id = fractional_id.first;
+        double fraction = fractional_id.second;
+        if (fraction >= resource_quantity) {
+          resource_ids.push_back(std::make_pair(resource_id, resource_quantity));
+          fractional_id.second -= resource_quantity;
+          found = true;
+          break;
+        }
+      }
+
+      if (!found) {
+        // In this case, there are not enough fractional resources, so we must
+        // use a whole ID.
+        RAY_CHECK(whole_ids.size() > 0);
+        int64_t resource_id = whole_ids.back();
+        whole_ids.pop_back();
+        resource_ids.push_back(std::make_pair(resource_id, resource_quantity));
+        fractional_ids.push_back(std::make_pair(resource_id, 1 - resource_quantity));
+      }
+    }
+
+  }
+  return acquired_resources;
+}
+
+void FractionalResourceIDSet::ReturnResources(const ResourceIdSet &resource_id_set) {
+
+  for (auto const &resource_pair : resource_id_set) {
+    auto resource_name = resource_pair.first;
+
+    auto it = resource_availabilities_.find(resource_name);
+    RAY_CHECK(it != resource_availabilities_.end());
+
+    auto &whole_ids = it->second.first;
+    auto &fractional_ids = it->second.second;
+
+    for (auto const &resource_info : resource_pair.second) {
+      int64_t resource_id = resource_info.first;
+      double resource_quantity = resource_info.second;
+      RAY_CHECK(resource_quantity > 0);
+
+      if (resource_quantity >= 1) {
+        // Handle the whole case.
+        int64_t whole_quantity = resource_quantity;
+        RAY_CHECK(whole_quantity == resource_quantity);
+
+        whole_ids.push_back(resource_id);
+
+      } else {
+        // Handle the fractional case.
+        bool found_id = false;
+        for (auto it = fractional_ids.begin(); it != fractional_ids.end(); ++it) {
+          if (it->first == resource_id) {
+            double sum = it->second + resource_quantity;
+            RAY_CHECK(sum <= 1);
+            if (sum < 1) {
+              it->second = sum;
+            } else {
+              // Return this ID to the whole IDs.
+              fractional_ids.erase(it);
+              whole_ids.push_back(resource_id);
+            }
+            found_id = true;
+            break;
+          }
+        }
+        RAY_CHECK(found_id);
```

Comment:
auto it = std::find_if(fractional_ids.begin(), fractional_ids.end(), [](auto& fractional_id) {return fractional_id.first == resource_id;});
RAY_CHECK(it !=  fractional_ids.end());

double sum = it->second + resource_quantity;
RAY_CHECK(sum <= 1);
 if (sum < 1) {
   it->second = sum;
} else {
   // Return this ID to the whole IDs.
  fractional_ids.erase(it);
  whole_ids.push_back(resource_id);
}

Reply:
That's much cleaner, thanks!